tests/org.eclipse.n4js.ui.tests/src/org/eclipse/n4js/tests/misc/ClearCacheOnCleanPluginUITest.java

```diff
@@ -0,0 +1,171 @@
+		super.setUp();
+		shippedCodeInitializeTestHelper.setupBuiltIns();
+	}
+
+	@After
+	@Override
+	public void tearDown() throws Exception {
+		// clear library manager to avoid confusing tests being executed after this test class
+		final MultiStatus multistatus = statusHelper
+				.createMultiStatus("Status of deleting NPM packages from library manager.");
+		externals.maintenanceDeleteNpms(multistatus);
+
+		shippedCodeInitializeTestHelper.tearDownBuiltIns();
+		super.tearDown();
+	}
+
+	/**
+	 * Tests if the {@link MultiCleartriggerCache} is cleared when a 'Clean Build' is triggered.
+	 */
+	@Test
+	public void testClearOnCleanBuild() throws CoreException {
+		File prjDir = new File(getResourceUri(PROBANDS, SUBFOLDER));
+		IProject project = ProjectTestsUtils.importProject(prjDir, PROJECT_NAME);
+		IResource resourceABC = project.findMember("src/ABC.n4js");
+		IFile fileABC = ResourcesPlugin.getWorkspace().getRoot().getFile(resourceABC.getFullPath());
+		assertTrue(fileABC.exists());
+		IResourcesSetupUtil.fullBuild();
+		waitForAutoBuild();
+
+		// use key of API_IMPL_MAPPING
+		SupplierWithPostAction testSupplier = new SupplierWithPostAction();
+		assertTrue(testSupplier.postActionTriggerCount == 0);
+
+		cache.clear(MultiCleartriggerCache.CACHE_KEY_API_IMPL_MAPPING);
+
+		// cache should be empty
+		Object test = cache.get(testSupplier, MultiCleartriggerCache.CACHE_KEY_API_IMPL_MAPPING);
+		assertTrue(testSupplier.postActionTriggerCount == 1);
+		assertTrue("test".equals(test));
+
+		// cache should contain key
+		cache.get(testSupplier, MultiCleartriggerCache.CACHE_KEY_API_IMPL_MAPPING);
+		assertTrue(testSupplier.postActionTriggerCount == 1);
+		assertTrue("test".equals(test));
+
+		// cleanBuild should clear the cache and set a new instance of ApiImplMappings
+		IResourcesSetupUtil.cleanBuild();
+		waitForAutoBuild();
+		test = cache.get(testSupplier, MultiCleartriggerCache.CACHE_KEY_API_IMPL_MAPPING);
+		assertTrue(test instanceof ApiImplMapping);
+	}
+
+	/**
+	 * Tests if the {@link MultiCleartriggerCache} is cleared when the user edits a 'package.json' file. Note that this
+	 * triggering is performed for the key {@link MultiCleartriggerCache#CACHE_KEY_API_IMPL_MAPPING} but not necessarily
+	 * for every other key in the cache.
+	 */
+	@Test
+	public void testClearOnModifyPackageJson() throws CoreException {
+		File prjDir = new File(getResourceUri(PROBANDS, SUBFOLDER));
+		IProject project = ProjectTestsUtils.importProject(prjDir, PROJECT_NAME);
+		IResource packagejson = project.findMember("package.json");
+		IFile filePJ = ResourcesPlugin.getWorkspace().getRoot().getFile(packagejson.getFullPath());
+		IResourcesSetupUtil.fullBuild();
+		waitForAutoBuild();
+
+		// use key of API_IMPL_MAPPING
+		SupplierWithPostAction testSupplier = new SupplierWithPostAction();
+		assertTrue(testSupplier.postActionTriggerCount == 0);
+
+		cache.clear(MultiCleartriggerCache.CACHE_KEY_API_IMPL_MAPPING);
+
+		// cache should be empty
```

Comment:
I don't understand this part. How can testing that cache is empty be done with the following code?

Reply:
Good question; when the method `get` is called it will invoke the supplier which will then increment the counter `postActionTriggerCount`. Hence, if the counter is equal to `1` it must have been zero before calling `get`. That means that the cache could not have contained the value of the supplier (before calling `get`).